### DISCUSSION TITLE
[suggestion] action without schema and typescript support.

### DISCUSSION THREAD

**Discussion by fitimbytyqi** [next-safe-action/next-safe-action#192] [Ideas]
This library is great 🔥, however, there are cases where people use NextJS as BFF including myself. In the majority of the cases we don't validate the inputs on the server because we already do it on the client & in the external server.

Is there any possibility to add types to the parsedInput (this will be whatever I've put in schema) or maybe have another property called input whose type we can modify ?

Most of our schemas use localization, the docs suggest to create a getSchema server action to pass the translations to that schema but on large applications it will be too messy to create a getSchema for each server action.

E.x Trpc lets us modify .schema's type without passing a zod schema by doing .schema(input => input as WhateverType).

Thanks for your time.

**Comment by adorkable77** [comment]
maybe you can do like this:
<img width="658" alt="image" src="https://github.com/TheEdoRan/next-safe-action/assets/28518813/16b13f4f-23db-4463-b96b-c384b57f1d28">

**Reply by fitimbytyqi** [reply]
@adorkable77 This is not working in my case 
![Screenshot 2024-07-05 at 10 46 06](https://github.com/TheEdoRan/next-safe-action/assets/61948517/5c43a046-39f6-4102-b649-991343d4c57a)

**Reply by adorkable77** [reply]
I tried this on the experimental version

**Reply by fitimbytyqi** [reply]
@adorkable77 Ideally, omitting schema and still be able to type the parameters is something I'm looking for.

**Reply by adorkable77** [reply]
@fitimbytyqi https://github.com/TheEdoRan/next-safe-action/issues/174#issuecomment-2175979834

**Reply by TheEdoRan** [reply]
It works for me too. You're probably using TS 5.5 and so you need to install `next-safe-action@experimental` to make it actually work, as @adorkable77 said.

You could probably write an helper function like this one to return `z.any()` as the input validator and just type the input:

![z.any input validator with typed input](https://github.com/TheEdoRan/next-safe-action/assets/1337629/5ce532cc-13cc-4fac-a3c0-d23a66e99db6)

**Reply by fitimbytyqi** [reply]
@TheEdoRan Thanks a lot for the helper function.

Updating next-safe action to it's experimental version solved the issue.

**Reply by TheEdoRan** [reply]
@fitimbytyqi Great! By the way, forgot to actually paste the code snippet with the function definition. Here it is:

```typescript
import { z } from "zod"

// Helper function that types the input but uses `z.any()` as input validator.
function typedAnySchema<S extends z.ZodType>() {
  return z.any() as unknown as S
}
```

**Reply by fitimbytyqi** [reply]
@TheEdoRan No biggie, had a bit fun writing it myself hahaha.

**Reply by MrOxMasTer** [reply]
> You could probably write an helper function like this one to return `z.any()` as the input validator and just type the input:

Maybe it's easy to do so that `clientInput` can be passed to `action` and `stateAction`, so as not to do so?

**Reply by MrOxMasTer** [reply]
> You could probably write an helper function like this one to return `z.any()` as the input validator and just type the input:

Or let people override the schema method. Instead of expanding the scheme, let people use its functions. The extension is not necessary, it can be done without this method. This will solve the problem with the resolver from react-hook-form and the ability to return any kind of verification

https://github.com/TheEdoRan/next-safe-action/discussions/209
https://github.com/TheEdoRan/next-safe-action/discussions/206
